### ISSUE TITLE
iOS Title property is not received when Alert property is a string

### ISSUE THREAD

**Issue by chadpav** [parse-community/parse-server-push-adapter#286]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server-push-adapter/security/policy).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/main/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server-push-adapter/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse Server Push Adapter](https://github.com/parse-community/parse-server-push-adapter/releases). 

### Issue Description
iOS devices were only receiving the alert body. The alert title was not showing up. This is the payload I tested with.

```
const result = await Parse.Push.send(
    {
      where: pushQuery,
      // WORKS FOR ANDROID
      notification: {
        title: 'Android title',
        body: 'Android notification body',
      },
      sound: 'default',
      // WORKS FOR IOS
      data: {
        // alert: { title: 'iOS data alert title', body: 'iOS data alert body' }, // <-- this works, avoids the bug in the adapter code, and sets both title+body
        title: 'iOS data title',
        alert: 'iOS data alert', // <-- will overwrite the title key
        priority: '10',
        collapse_id: '1',
      },
    },
    {
      useMasterKey: true,
    }
  )
```

### Steps to reproduce
1. Configure project using FCM to send iOS push notifications.
2. Send notification to an iOS device with the payload provided above.

### Actual Outcome
- Push notification is received, however the title was missing.

### Expected Outcome
- Should have received both title and alert message.

###  Environment
- `"@parse/push-adapter": "^6.5.0"`
- `"parse-server": "~7.2.0"`

### Logs
- not relevant, logs stated that push notification was sent successfully.

**Comment by chadpav** [comment]
I stepped into the push-adapter code and figured out the issue. The code was initializing the `apns.payload.aps.alert` key before setting the `body` key. This is a code snippet from `FCM.js` in `_APNStoFCMPayload()`:
```
    case 'alert':
      if (typeof coreData.alert == 'object') {
        // When we receive a dictionary, use as is to remain
        // compatible with how the APNS.js + node-apn work
        apnsPayload['apns']['payload']['aps']['alert'] = coreData.alert;
      } else {
        // When we receive a value, prepare `alert` dictionary
        // and set its `body` property
        apnsPayload['apns']['payload']['aps']['alert'] = {};
        apnsPayload['apns']['payload']['aps']['alert']['body'] = coreData.alert;
      }
      break;
```

I learned through reading the code that I could send both alert/title as an object as a workaround.

I'm submitting a PR with a failing test + patch.

**Comment by chadpav** [comment]
You are correct that it wasn't working when passing alert as a string instead of an object. I just lost an afternoon troubleshooting why it wasn't working. I did discover the alternate method by reading the adapter source code but felt that I could save someone else the pain by providing a simple fix.

It was not obvious what payload the adapter is expecting in order to send notifications to iOS and Android devices. Maybe I'll provide a more complete code sample in the readme with my PR.

**Comment by mman** [comment]
@chadpav The point is that currently supported syntax is the one officially supported by Apple APNS, and only passed one to one via FCM to APNS.

The one you are proposing is artificial new syntax that is not expected by anybody...

am I missing something?

**Comment by chadpav** [comment]
ok, I hear what you are saying. Let me check the Apple APNS docs and get back to you. On the surface it makes no sense to support setting `title` and `alert` as a string and then silently eating the `title` key. It looks like a bug but if your point is that is just have APNS works then I understand your point.

If I have to study the Parse docs, FCM docs, and APNS docs in order to work out what the correct payload is then maybe we could provide more code samples or even a validator to give feedback?

**Comment by mman** [comment]
Yeah, I have been there as well multiple times.

First came iOS with some simple syntax, then they amended the syntax to support title, subtitle, and then they were adding more and more. My favourite was `content-available: 1` where you can not pass anything else but number otherwise stuff breaks... we have even tests for this. 

Then came Android with their own GCM syntax... and Parse Push adapter added code to send iOS formatted pushes properly to Android as well via GCM... while skipping unsupported fields. Then came FCM with their own syntax and code was added to transform it to iOS payload, or Android payload... depending on destination. and this is where we are now.

So it is absolutely possible to send one payload to multiple destinations regardless of what adapter you talk to - once you stick to currently supported payload format - but what is missing is proper documentation so that you do not need to lookup the code, and you do not hit obstacles as you did.

Adding more potentially valid input variations will only make "adding the proper documentation" task more complex...

So I think what is needed is to add more detailed documentation of the currently supported payload format while documenting all caveats and features supported on all platforms... and/or possibly add a validator that may catch invalid payload during runtime... that would definitely help...

Otherwise we end up in a state where we have

1. APNS payload syntax
2. FCM payload syntax
3. Parse Server Push Adapter payload syntax

with 1. and 2. being used in the wild with no easy/clear path to migrate towards 3. So we will have to support 1. and 2. indefinitely anyway...

so +1 for docs, and +1 for validator 

just my .2 cents,
Martin

**Comment by chadpav** [comment]
@mman following up with what the Parse Server guide says... what I was trying to do is [documented here](https://docs.parseplatform.org/parse-server/guide/#4-send-push-notifications). Although that documentation is really light and old.

The `parse-server-push-adapter` doesn't cover any of this syntax or what is supported. It only covers how to configure the adapter.
```
curl -X POST \
  -H "X-Parse-Application-Id: you_app_id" \
  -H "X-Parse-Master-Key: your_master_key" \
  -H "Content-Type: application/json" \
  -d '{
        "where": {
          "deviceType": {
            "$in": [
              "ios",
              "android"
            ]
          }
        },
        "data": {
          "title": "The Shining",
          "alert": "All work and no play makes Jack a dull boy."
        }
      }'\   http://your_server_address/parse/push
```

**Comment by mman** [comment]
@chadpav Excellent catch, exactly as you say... this is one of the oldest docs that was not updated in ages... and more docs missing... although existence of this example may prove that it actually was supposed to work this way at some point in the past...

**Comment by chadpav** [comment]
@mman I can see where this slowly got out of hand over the years. 

Lets decide what we should do with this issue/PR. Are there others that should weigh in? I see some options, we could close the issue and document the "alert object" syntax here. I almost hate to add a code sample in the readme because I'm not testing every supported configuration. Or we could merge this fix since it's still backward compatible with existing clients.

I see another, parallel option forward that addresses issue #245. Fork this repo, rename it to `parse-server-fcm-push-adapter`, remove all other configuration options, add support for the full FCM syntax, add a link in the readme pointing to the FCM documentation that says "that's the api". It's clean and doesn't affect existing users.

**Comment by mman** [comment]
I'll let @mtrezza to chime in what he sees as the best approach. I think the Adapter interface should document and support same interface no matter what the underlying implementation may be... so making different push adapters support different syntax of payload is a NO GO for me... as you can not swap them easily... but there may be a precedent that I do not see where different adapter implementations may support different features... (Mongo versus PostgreSQL comes to mind)...

I honestly kind of like the validator approach, and merging your PR to make the existing example work... seems like a good way forward...

**Comment by chadpav** [comment]
I think that means that Parse Server should have an opinionated, documented API that the adapter translates to it's destination.

I'd also argue that the validation should happen in Parse Server and not each individual adapter if you want to make them truly swappable.

**Comment by mtrezza** [comment]
Interesting discussion, we have been here before. I think we want to distinguish between 2 types of APIs:

- a) Parse push adapter API:
  - Syntax is designed and documented by Parse
  - Payload is modified by the adapter as needed before being sent to the push provider.
  - Changes influence the module version (e.g. breaking changes)
  - Purpose is to make life easier for developers by providing a unified syntax across push providers.
  - The API is our responsibility to manage.

- b) 3rd party APIs where the adapter just passes through the payload without (much) validation. All the points above do not apply here.

**Comment by jimnor0xF** [comment]
Have not read through the entire discussion here, but you can access the raw FCM API without any abstractions by putting it inside a key `rawPayload`. 

Example:
```javascript
  const q = new Parse.Query(Parse.Installation);
  q.containedIn('user', userIds);

  await Parse.Push.send(
    {
      where: q,
      rawPayload: {
        notification: {
          title: title,
          body: body,
        },
        data: {
          type: type,
          payload: payload,
        },
        android: {
          priority: 'high',
        },
        apns: {
          headers: {
            'apns-priority': '5',
          },
          payload: {
            aps: {
              contentAvailable: true,
            },
          },
        },
      },
    },
    { useMasterKey: true },
  );
```

**Comment by jimnor0xF** [comment]
@chadpav 
Have not read through the entire discussion 
here, but you can access the raw FCM API without any abstractions by putting it inside a key `rawPayload`. 

Example:
```javascript
  const q = new Parse.Query(Parse.Installation);
  q.containedIn('user', userIds);

  await Parse.Push.send(
    {
      where: q,
      rawPayload: {
        notification: {
          title: title,
          body: body,
        },
        data: {
          type: type,
          payload: payload,
        },
        android: {
          priority: 'high',
        },
        apns: {
          headers: {
            'apns-priority': '5',
          },
          payload: {
            aps: {
              contentAvailable: true,
            },
          },
        },
      },
    },
    { useMasterKey: true },
  );
```

**Comment by chadpav** [comment]
@jimnor0xF I was wondering about that key but never looked into it. That is what I'll probably end up doing. Thanks!

**Comment by mtrezza** [comment]
How can we resolve this issue in the sense of https://github.com/parse-community/parse-server-push-adapter/issues/286#issuecomment-2289294545? Update the docs?

**Comment by chadpav** [comment]
I think you are right that we should document the payload that Parse Server supports. If we include the rawPayload in the docs then it unblocks most people. While I still believe the fix that I provided is a legit bug, I can agree with the logic that we shouldn't go changing things when it's not the intended syntax to begin with.

The only issue that I see is that I'm not certain that Parse Server is even consistent with its API. By that I mean that the API might be different depending on how you configure the push adapter that ships with Parse Server (e.g., APNS vs FCM configuration for iOS devices). I could be wrong on that.

Is there anything you need from me?

**Comment by chadpav** [comment]
@mtrezza side note but I'm running the latest Parse Dashboard and when I send messages to my Android device they don't display to the user. I need to step into the code to see if the Parse Dashboard is using the correct payload. I'll get back to you on that.

**Comment by mtrezza** [comment]
It's actually a good time to clean this up now. We're just ~4 months from Parse Server 8 release where we can introduce breaking changes.

If there is a legacy syntax that we want to remove, then let's prepare a PR for that. We'll merge it in Nov as a breaking change and then bundle it with Parse Server 8.

We can already update the docs now.

**Comment by jimnor0xF** [comment]
> Interesting discussion, we have been here before. I think we want to distinguish between 2 types of APIs:
> 
> - a) Parse push adapter API:
>   - Syntax is designed and documented by Parse
>   - Payload is modified by the adapter as needed before being sent to the push provider.
>   - Changes influence the module version (e.g. breaking changes)
>   - Purpose is to make life easier for developers by providing a unified syntax across push providers.
>   - The API is our responsibility to manage.
> 
> - b) 3rd party APIs where the adapter just passes through the payload without (much) validation. All the points above do not apply here.
> 
> Maybe we just need to define which of the current APIs (or adapter configs / syntax parts) fall into which of the 2 categories above. And then polish the docs a bit.
> 
> Note that any breaking changes in the adapter will not be merged until end of the year, because Parse Server comes bundled with the push adapter, and we would not be able to upgrade it there if it contained a breaking change.

In the long term, if breaking changes are allowed, I'd prefer to not do a) at all and only do b).

a) takes a fair amount of maintenance and keeping a consistent standard without letting it rot over time is difficult.

Easier to just link to the push provider docs for the payload needed.

For now I think updating the docs a bit more is a good idea.

**Comment by mtrezza** [comment]
> I'd prefer to not do a) at all and only do b)

For people using providers directly (instead of FCM as an intermediary for all other providers), this would require every developer to recreate a logic for a universal internal API. Not sure how much sense it would make for us to remove that logic and roll the burden over to every individual developer, doesn't sound very efficient.

**Comment by mtrezza** [comment]
How should we proceed here?

**Comment by jimnor0xF** [comment]
@mtrezza 

> For developers using providers directly (instead of FCM as an intermediary for all other providers), this would require them to recreate the same logic for a universal internal API.

For me personally at least it would still be interesting if you could simply specify the exact payload for each provider you'd want to use. Sort of a "thin wrapper" approach.

Something like:
```
{
  "apns": {
   "key": value",
  },
  "fcm": {
    "key": "value",
  },
  "onesignal": {
    "key": "value",
  }
}

```

I admit this is more verbose than providing an abstracted payload. But, also means that less maintenance is needed since you don't need updates when a provider adds new keys or to extend what is currently possible with the abstracted payload. Simply having to link the docs for the provider on what payload is expected is enough.

You'd need to be able to map between device type and when to use which provider, and have this configurable by the server admin on the backend of course. When it's an android device, use fcm, when apple device use apns, default use this and so on.

But yeah, anyway, just my opinion on things. In practice, not sure how doable above proposal would be while also avoiding breaking changes. I'm pretty happy with how things are currently since I just use FCM for everything.

**Comment by mtrezza** [comment]
I don't see a raw provider API and an abstracted API as mutually exclusive. They can live side by side. The abstracted API can cover the common denominator across all push APIs. The raw API can be used for provider specific flavors.

To your point of "payload conflicts", there should be none. Maybe that's a separate issue?

**Comment by jimnor0xF** [comment]
@mtrezza 
I think having a raw provider API + abstracted API is probably the way to go then. Satisfies both sides.

We would need to standardize what the abstracted API should look like if we're going forward with this. Since currently it's a bit unclear to me whether the current one that is used for APNS/FCM is good or not.

>To your point of "payload conflicts", there should be none. Maybe that's a separate issue?

Yeah, not really related to our discussion. More so related to how things are handled in the code currently.
For example [here](https://github.com/parse-community/parse-server-push-adapter/blob/master/src/FCM.js#L270), we actually delete APNS-related keys if present when sending via FCM for android. If you have WEB configured as a provider in tandem, it will also send APNS/FCM related keys to WEB if you plan on having one single payload that fits all providers.

What we probably should do instead is to have a handler that converts the abstracted payload into whatever raw payload the configured provider(s) need before it is sent to the module that deals with the sending part. That way you only send what is necessary according to the provider.

**Comment by mtrezza** [comment]
We could use a model like the parse-server-api-mail-adapter where a internal converter is used to convert the generic payload to a provider specific payload. That creates a clear payload pipeline. I don't think this adapter is far from it.

**Comment by jimnor0xF** [comment]
@mtrezza 
Looks like a clean way of doing things.  Will see if I have some time over to look into this after the holidays perhaps.
